### PR TITLE
[Type checker] Reduce/isolate reuse of parent generic environments

### PR DESCRIPTION
**Pull request by DougGregor**
The type checker is still reusing the archetypes from parent contexts in nested generic contexts, which is is both a mess (semantically) and means that we cannot add any new requirements to those archetypes, a source of numerous bugs. Take a few more concrete steps toward eliminating this reuse of parent archetypes:

* Set the contextual type of the implicit `self` parameter late, after we've had a change to create the generic environment for a function
* Type-check the definitions of `typealias` declarations with the `typealias` as the context, so we're in the right generic environment for generic `typealias` declarations
* Eliminate `PartialGenericTypeToArchetypeResolver`, which handled the weird case of pulling archetypes from an outer context rather than the current context. Instead, be explicit about the places where we reuse archetypes by explicitly telling the archetype builder to reuse the parent archetypes.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

File: lib/Sema/TypeCheckDecl.cpp [hunk 1]
```diff
@@ -6820,6 +6833,25 @@ void TypeChecker::checkDeclCircularity(NominalTypeDecl *decl) {
   }
 }
 
+/// Determine whether the given variable is the implicit 'self' parameter for
+/// a function, and return that function if so.
+static AbstractFunctionDecl *isImplicitSelfParam(VarDecl *var){
```
Comment (slavapestov) [inline lib/Sema/TypeCheckDecl.cpp]: We already have VarDecl::isSelfParameter()...

File: lib/AST/DeclContext.cpp [hunk 2]
```diff
@@ -357,6 +357,12 @@ AbstractFunctionDecl *DeclContext::getInnermostMethodContext() {
   }
 }
 
+bool DeclContext::isTypeContext() const {
+  return (getContextKind() == DeclContextKind::GenericTypeDecl &&
```
Comment (slavapestov) [inline lib/AST/DeclContext.cpp]: maybe just isa<NominalTypeDecl>(this) then...

File: lib/Sema/ITCDecl.cpp [hunk 3]
```diff
@@ -103,7 +103,7 @@ void IterativeTypeChecker::processResolveInheritedClauseEntry(
 
   // Validate the type of this inherited clause entry.
   // FIXME: Recursion into existing type checker.
-  PartialGenericTypeToArchetypeResolver resolver;
+  GenericTypeToArchetypeResolver resolver(dc->getGenericEnvironmentOfContext());
```
Comment (slavapestov) [inline lib/Sema/ITCDecl.cpp]: There's a constructor that takes a dc too, it's nicer

File: lib/Sema/ITCDecl.cpp [hunk 4]
```diff
@@ -314,9 +314,10 @@ void IterativeTypeChecker::processResolveTypeDecl(
 
       // Note: recursion into old type checker is okay when passing in an
       // unsatisfied-dependency callback.
+      GenericTypeToArchetypeResolver resolver(typeAliasDecl);
       if (TC.validateType(typeAliasDecl->getUnderlyingTypeLoc(),
                           typeAliasDecl->getDeclContext(),
-                          options, nullptr, &unsatisfiedDependency)) {
+                          options, &resolver, &unsatisfiedDependency)) {
```
Comment (slavapestov) [inline lib/Sema/ITCDecl.cpp]: If the parameter is always non-null, it could be a reference

File: test/decl/ext/generic.swift [hunk 5]
```diff
@@ -151,4 +151,4 @@ struct S4<Q>: P4 {
   init(_: Q) { }
 }
 
-extension S4 where T : P5 {} // expected-error{{type 'T' in conformance requirement does not refer to a generic parameter or associated type}}
+extension S4 where T : P5 {}
```
Comment (slavapestov) [inline test/decl/ext/generic.swift]: Heh

File: lib/Sema/TypeCheckDecl.cpp [hunk 6]
```diff
@@ -732,11 +732,18 @@ TypeChecker::handleSILGenericParams(GenericParamList *genericParams,
     revertGenericParamList(genericParams);
 
     ArchetypeBuilder builder(*DC->getParentModule());
+    CompleteGenericTypeResolver completeResolver(*this, builder);
     checkGenericParamList(&builder, genericParams, parentSig, parentEnv,
-                          nullptr);
```
Comment (slavapestov) [inline lib/Sema/TypeCheckDecl.cpp]: reference instead of pointer maybe?

File: lib/Sema/TypeCheckDecl.cpp (hunk 1 above)
Comment (DougGregor) [inline lib/Sema/TypeCheckDecl.cpp]: Ugh! I was looking for that on `ParamDecl`. Thank you

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
Linux build failed with something unrelated in LLVM:

buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/llvm/lib/Target/PowerPC/PPCMCInstLower.cpp:40:22: error: no member named 'Mang' in 'llvm::AsmPrinter'
  Mangler *Mang = AP.Mang;

macOS build failed with an exception

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test macOS

**Comment by DougGregor** [comment]
@swift-ci please smoke test Linux

**Pull request merged**